networkx/linalg/algebraicconnectivity.py

```diff
@@ -244,7 +244,7 @@ def P(x):
             W -= (W.T * X * X.T).T
             project(W)
             # Compute the diagonal of P * L * P as a Jacobi preconditioner.
-            D = L.diagonal()
+            D = L.diagonal().astype(float)
             D += 2. * (asarray(X) * asarray(W)).sum(axis=1)
```

Comment:
It looks to me like the problem is the ```+=```, not the fact that we are adding ints to floats.
So we can keep the 3 lines as they are now, remove the ```D +=``` in the last two lies and add ```+``` at the end of the first two lines. It will be no harder to read than it is now (and easier to read in subtle ways like casting).

Is there a reason for the in-place addition here?